8293198: [vectorapi] Improve the implementation of VectorMask.indexInRange()

XiaohongGong opened the pull request:
The Vector API `"indexInRange(int offset, int limit)"` is used
to compute a vector mask whose lanes are set to true if the
index of the lane is inside the range specified by the `"offset"`
and `"limit"` arguments, otherwise the lanes are set to false.

There are two special cases for this API:
 1) If `"offset >= 0 && offset >= limit"`, all the lanes of the
generated mask are false.
 2) If` "offset >= 0 && limit - offset >= vlength"`, all the
lanes of the generated mask are true. Note that `"vlength"` is
the number of vector lanes.

For such special cases, we can simply use `"maskAll(false|true)"`
to implement the API. Otherwise, the original comparison with
`"iota" `vector is needed. And for further optimization, we have
optimal instruction supported by SVE (i.e. whilelo [1]), which
can implement the API directly if the `"offset >= 0"`.

As a summary, to optimize the API, we can use the if-else branches
to handle the specific cases in java level and intrinsify the
remaining case by C2 compiler:

```
  public VectorMask<E> indexInRange(int offset, int limit) {
      if (offset < 0) {
          return this.and(indexInRange0Helper(offset, limit));
      } else if (offset >= limit) {
          return this.and(vectorSpecies().maskAll(false));
      } else if (limit - offset >= length()) {
          return this.and(vectorSpecies().maskAll(true));
      }
      return this.and(indexInRange0(offset, limit));
 }
```

The last part (i.e. `"indexInRange0"`) in the above implementation
is expected to be intrinsified by C2 compiler if the necessary IRs
are supported. Otherwise, it will fall back to the original API
implementation (i.e. `"indexInRange0Helper"`). Regarding to the
intrinsifaction, the compiler will generate `"VectorMaskGen"` IR
with "limit - offset" as the input if the current platform supports
it. Otherwise, it generates `"VectorLoadConst + VectorMaskCmp"` based
on `"iota < limit - offset"`.

For the following java code which uses `"indexInRange"`:

```
static final VectorSpecies<Double> SPECIES =
                                   DoubleVector.SPECIES_PREFERRED;
static final int LENGTH = 1027;

public static double[] da;
public static double[] db;
public static double[] dc;

private static void func() {
    for (int i = 0; i < LENGTH; i += SPECIES.length()) {
        var m = SPECIES.indexInRange(i, LENGTH);
        var av = DoubleVector.fromArray(SPECIES, da, i, m);
        av.lanewise(VectorOperators.NEG).intoArray(dc, i, m);
    }
}
```

The core code generated with SVE 256-bit vector size is:

```
  ptrue   p2.d                  ; maskAll(true)
  ...
LOOP:
  ...
  sub     w11, w13, w14         ; limit - offset
  cmp     w14, w13
  b.cs    LABEL-1               ; if (offset >= limit) => uncommon-trap
  cmp     w11, #0x4
  b.lt    LABEL-2               ; if (limit - offset < vlength)
  mov     p1.b, p2.b
LABEL-3:
  ld1d    {z16.d}, p1/z, [x10]  ; load vector masked
  ...
  cmp     w14, w29
  b.cc    LOOP
  ...
LABEL-2:
  whilelo p1.d, x16, x10        ; VectorMaskGen
  ...
  b       LABEL-3
  ...
LABEL-1:
  uncommon-trap
```

Please note that if the array size `LENGTH` is aligned with
the vector size 256 (i.e. `LENGTH = 1024`), the branch "LABEL-2"
will be optimized out by compiler and it becomes another
uncommon-trap.

For NEON, the main CFG is the same with above. But the compiler
intrinsification is different. Here is the code:

```
  sub     x10, x10, x12          ; limit - offset
  scvtf   d16, x10
  dup     v16.2d, v16.d[0]       ; replicateD

  mov     x8, #0xd8d0
  movk    x8, #0x84cb, lsl #16
  movk    x8, #0xffff, lsl #32
  ldr     q17, [x8], #0          ; load the "iota" const vector
  fcmgt   v18.2d, v16.2d, v17.2d ; mask = iota < limit - offset
```

Here is the performance data of the new added benchmark on an ARM
SVE 256-bit platform:

```
Benchmark                               (size)  Before    After   Units
IndexInRangeBenchmark.byteIndexInRange   1024 11203.697 41404.431 ops/ms
IndexInRangeBenchmark.byteIndexInRange   1027  2365.920  8747.004 ops/ms
IndexInRangeBenchmark.doubleIndexInRange 1024  1227.505  6092.194 ops/ms
IndexInRangeBenchmark.doubleIndexInRange 1027   351.215  1156.683 ops/ms
IndexInRangeBenchmark.floatIndexInRange  1024  1468.876 11032.580 ops/ms
IndexInRangeBenchmark.floatIndexInRange  1027   699.645  2439.671 ops/ms
IndexInRangeBenchmark.intIndexInRange    1024  2842.187 11903.544 ops/ms
IndexInRangeBenchmark.intIndexInRange    1027   689.866  2547.424 ops/ms
IndexInRangeBenchmark.longIndexInRange   1024  1394.135  5902.973 ops/ms
IndexInRangeBenchmark.longIndexInRange   1027   355.621  1189.458 ops/ms
IndexInRangeBenchmark.shortIndexInRange  1024  5521.468 21578.340 ops/ms
IndexInRangeBenchmark.shortIndexInRange  1027  1264.816  4640.504 ops/ms

```
And the performance data with ARM NEON:

```
Benchmark                               (size)  Before    After   Units
IndexInRangeBenchmark.byteIndexInRange   1024  4026.548 15562.880 ops/ms
IndexInRangeBenchmark.byteIndexInRange   1027   305.314   576.559 ops/ms
IndexInRangeBenchmark.doubleIndexInRange 1024   289.224  2244.080 ops/ms
IndexInRangeBenchmark.doubleIndexInRange 1027    39.740    76.499 ops/ms
IndexInRangeBenchmark.floatIndexInRange  1024   675.264  4457.470 ops/ms
IndexInRangeBenchmark.floatIndexInRange  1027    79.918   144.952 ops/ms
IndexInRangeBenchmark.intIndexInRange    1024   740.139  4014.583 ops/ms
IndexInRangeBenchmark.intIndexInRange    1027    78.608   147.903 ops/ms
IndexInRangeBenchmark.longIndexInRange   1024   400.683  2209.551 ops/ms
IndexInRangeBenchmark.longIndexInRange   1027    41.146    69.599 ops/ms
IndexInRangeBenchmark.shortIndexInRange  1024  1821.736  8153.546 ops/ms
IndexInRangeBenchmark.shortIndexInRange  1027   158.810   243.205 ops/ms

```
The performance improves about `3.5x ~ 7.5x` on the vector size aligned
(1024 size) benchmarks both with NEON and SVE. And it improves about
`3.5x/1.8x` on the vector size not aligned (1027 size) benchmarks with
SVE/NEON respectively. We can also observe the similar improvement on
the x86 platforms.

[1] https://developer.arm.com/documentation/ddi0596/2020-12/SVE-Instructions/WHILELO--While-incrementing-unsigned-scalar-lower-than-scalar-

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293198](https://bugs.openjdk.org/browse/JDK-8293198): [vectorapi] Improve the implementation of VectorMask.indexInRange()


### Reviewers
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**) ⚠️ Review applies to [6f7c30be](https://git.openjdk.org/jdk/pull/12064/files/6f7c30be545442b0be18c1567580ac7b79a5380a)
 * [Quan Anh Mai](https://openjdk.org/census#qamai) (@merykitty - Committer) ⚠️ Review applies to [9388e29d](https://git.openjdk.org/jdk/pull/12064/files/9388e29d7bcb1bbf136cb64e548bbd0e681e811d)
 * [Paul Sandoz](https://openjdk.org/census#psandoz) (@PaulSandoz - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12064/head:pull/12064` \
`$ git checkout pull/12064`

Update a local copy of the PR: \
`$ git checkout pull/12064` \
`$ git pull https://git.openjdk.org/jdk pull/12064/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12064`

View PR using the GUI difftool: \
`$ git pr show -t 12064`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12064.diff">https://git.openjdk.org/jdk/pull/12064.diff</a>

</details>

merykitty commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/AbstractMask.java:
```diff
@@ -195,24 +195,52 @@ static long toLongHelper(boolean[] bits) {
+    @Override
+    @ForceInline
+    public VectorMask<E> indexInRange(int offset, int limit) {
+        if (offset < 0) {
```
These fast-paths penalise every usage of `VectorMask::indexInRange`, especially the common use case of tail processing an array. So I don't think it is needed, the user can implement it themselves if their use cases find it beneficial. Thanks.

XiaohongGong replied:
This "offset < 0" path can be optimized out by compiler if the "offset >= 0" in the common cases (i.e. normal loop with no tail loop).

merykitty replied:
Yes but I don't think it is common enough to deserve special treatment. A fast path should be common for this branch otherwise we are regressing the common path for the corner cases.

XiaohongGong replied:
The special handling here is necessary for me, because we do not have the instruction supports in compiler intrinsification. Besides, the original API implementation also has the special handle for "offset < 0", which means this API itself accepts the negative offset as the valid argument.  We have to conver it here as well to make the implementation comprehensive.

 Maybe I misunderstood what you mean. Could you please give an alternative solution besides  the current implementation for such corner cases? Thanks in advance!

XiaohongGong commented:
Hi @PaulSandoz , @jatin-bhateja, @sviswa7, could you please help to take a look at this optimization? Any feedback is welcome. Thanks in advance!

jatin-bhateja commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/AbstractMask.java:
```diff
@@ -195,24 +195,52 @@ static long toLongHelper(boolean[] bits) {
+            return this.and(indexInRange0Helper(offset, limit));
+        } else if (offset >= limit) {
+            return vectorSpecies().maskAll(false);
+        } else if (limit - offset >= length()) {
```
Your patch re-organized the code to take care of following cases :-
1.   offset < 0                         : Still falling over to old code.
2.   offset >= limit                  : Special case optimized 
3.   (limit - offset) >= length  :  Most general case.
4.   Partial in range check       :  New intrinsic, existing java side implementation in checkIndex0 is also composed of intrinsified calls,  infrequently taken paths will anyways lead to uncommon traps.

Have you tried introducing just case 3 (first) and then case 2 in existing indexInRange implementation.

XiaohongGong replied:
Thanks for the review @jatin-bhateja !

> Have you tried introducing just case 3 (first) and then case 2 in existing indexInRange implementation.

Yes, I tried with this way as well, together with the performance testing compared with current version. The result is:
1. For cases that the array size is aligned with the vector size, the performance is almost the same with each other.
2. For cases that the array size is not aligned (which needs the masked operation), the performance with current version can improve about 5%~6% compared with the way as you suggested. But the improvement is limited to the cases with smaller vector size, whose tail loop size is bigger.

So, adding the new intrinsic has no side effect for me, except for the more complex java side code and one additional intrinsic.

XiaohongGong replied:
> Can you move this else if check at the top, this is the most general case appearing in the loop and hence two extra uncommon trap jumps before it for special cases may penalize this.

The generated code does not have any difference by moving the most general case at the top. It still will generate the uncommon trap jumps before it. Besides, moving it at the top may make the code in java side harder to understand. The code may seem like:
```
public VectorMask<E> indexInRange(int offset, int limit) {
        if (offset > 0 && limit - offset >= length()) {
            return this;
        } else if (offset > 0 && offset < limit) {
            return this.and(indexInRange0(offset, limit));
        } else if (offset >= limit) {
            return vectorSpecies().maskAll(false);
        } else {
           return this.and(indexInRange0Helper(offset, limit));
        }
}
``` 

But as what you suggested above, if we removing the new intrinsic and keep using the "checkIndex0" to handle the tail loop, the code may be easier, such as:
```
public VectorMask<E> indexInRange(int offset, int limit) {
        if (offset > 0 && limit - offset >= length()) {
            return this;
        } else if (offset >= limit) {
            return vectorSpecies().maskAll(false);
        } else {
           return this.and(indexInRange0Helper(offset, limit));
        }
}
```

XiaohongGong replied:
> Thanks for the review @jatin-bhateja !
> 
> > Have you tried introducing just case 3 (first) and then case 2 in existing indexInRange implementation.
> 
> Yes, I tried with this way as well, together with the performance testing compared with current version. The result is:
> 
> 1. For cases that the array size is aligned with the vector size, the performance is almost the same with each other.
> 2. For cases that the array size is not aligned (which needs the masked operation), the performance with current version can improve about 5%~6% compared with the way as you suggested. But the improvement is limited to the cases with smaller vector size, whose tail loop size is bigger.
> 
> So, adding the new intrinsic has no side effect for me, except for the more complex java side code and one additional intrinsic.

Another benefit is the code change will be simpler, that almost all the files except the `AbstractMask.java` do not need to be changed. I will make the modification and rerun the benchmarks to decide whether it's necessary to add the new intrinsics. Thanks!

jatin-bhateja replied:
While you talked about Java side changes, I found another opportunity for optimization in checkIndex0 implementation, in the following code snippet from checkIndex0 method, indexLimit is guaranteed to be a +ve value.

```
         int indexLimit = Math.max(0, Math.min(length - offset, vlength));
         VectorMask<E> badMask =
            iota.compare(GE, iota.broadcast(indexLimit)); 
```
For float/double iota vectors, subsequent broadcast operation accepting long argument checks for precision loss before broadcasting floating point value.  

```
       long longToElementBits(long value) {
            // Do the conversion, and then test it for failure.
            float e = (float) value;
            if ((long) e != value) {
                throw badElementBits(value, e);
            }
            return toBits(e);
        }`
```
 
This can be saved by doing a prior casting of indexLimit to floating point value before calling broadcast, effectively we may need to move checkIndex0 to template generated abstract vector classes.

XiaohongGong replied:
Hi, I modified a version by using the old implementation for the tail loop instead of adding the new intrinsics. The code looks like:
```
public VectorMask<E> indexInRange(int offset, int limit) {
        int vlength = length();
        if (offset >= 0 && limit - offset >= length()) {
            return this;
        } else if (offset >= limit) {
            return vectorSpecies().maskAll(false);
        }

        Vector<E> iota = vectorSpecies().zero().addIndex(1);
        VectorMask<E> badMask = checkIndex0(offset, limit, iota, vlength);
        return this.andNot(badMask);
    }
```
And I tested the performance of the new added benchmarks with different vector size on NEON/SVE and x86 avx2/avx512 architectures. The results show that the performance of changed version is worse than the current version, if the array size is not aligned with the vector size.

Here are some raw data with NEON:
```
Benchmark                                 (size)   Mode  Cnt  current   modified   Units
IndexInRangeBenchmark.byteIndexInRange       131  thrpt    5  2654.919  2584.423   ops/ms
IndexInRangeBenchmark.byteIndexInRange       259  thrpt    5  1830.364  1802.876   ops/ms
IndexInRangeBenchmark.byteIndexInRange       515  thrpt    5  1058.548  1073.742   ops/ms
IndexInRangeBenchmark.doubleIndexInRange     131  thrpt    5   594.920   593.832   ops/ms
IndexInRangeBenchmark.doubleIndexInRange     259  thrpt    5   308.678   149.279   ops/ms
IndexInRangeBenchmark.doubleIndexInRange     515  thrpt    5   160.639    74.579   ops/ms
IndexInRangeBenchmark.floatIndexInRange      131  thrpt    5  1097.567  1104.008   ops/ms
IndexInRangeBenchmark.floatIndexInRange      259  thrpt    5   617.845   606.886   ops/ms
IndexInRangeBenchmark.floatIndexInRange      515  thrpt    5   315.978   152.046   ops/ms
IndexInRangeBenchmark.intIndexInRange        131  thrpt    5  1165.279  1205.486   ops/ms
IndexInRangeBenchmark.intIndexInRange        259  thrpt    5   633.648   631.672   ops/ms
IndexInRangeBenchmark.intIndexInRange        515  thrpt    5   315.370   154.144   ops/ms
IndexInRangeBenchmark.longIndexInRange       131  thrpt    5   639.840   633.623   ops/ms
IndexInRangeBenchmark.longIndexInRange       259  thrpt    5   312.267   152.788   ops/ms
IndexInRangeBenchmark.longIndexInRange       515  thrpt    5   163.028    78.150   ops/ms
IndexInRangeBenchmark.shortIndexInRange      131  thrpt    5  1834.318  1800.318   ops/ms
IndexInRangeBenchmark.shortIndexInRange      259  thrpt    5  1105.695  1094.347   ops/ms
IndexInRangeBenchmark.shortIndexInRange      515  thrpt    5   602.442   599.827   ops/ms
```

And the data with SVE 256-bit vector size:
```
Benchmark                                 (size)   Mode  Cnt   current   modified Units
IndexInRangeBenchmark.byteIndexInRange       131  thrpt    5  23772.370 22921.113 ops/ms
IndexInRangeBenchmark.byteIndexInRange       259  thrpt    5  18930.388 17920.910 ops/ms
IndexInRangeBenchmark.byteIndexInRange       515  thrpt    5  13528.610 13282.504 ops/ms
IndexInRangeBenchmark.doubleIndexInRange     131  thrpt    5   7850.522  7975.720 ops/ms
IndexInRangeBenchmark.doubleIndexInRange     259  thrpt    5   4281.749  4373.926 ops/ms
IndexInRangeBenchmark.doubleIndexInRange     515  thrpt    5   2160.001   604.458 ops/ms
IndexInRangeBenchmark.floatIndexInRange      131  thrpt    5  13594.943 13306.904 ops/ms
IndexInRangeBenchmark.floatIndexInRange      259  thrpt    5   8163.134  7912.343 ops/ms
IndexInRangeBenchmark.floatIndexInRange      515  thrpt    5   4335.529  4198.555 ops/ms
IndexInRangeBenchmark.intIndexInRange        131  thrpt    5  22106.880 20348.266 ops/ms
IndexInRangeBenchmark.intIndexInRange        259  thrpt    5  11711.588 10958.299 ops/ms
IndexInRangeBenchmark.intIndexInRange        515  thrpt    5   5501.034  5358.441 ops/ms
IndexInRangeBenchmark.longIndexInRange       131  thrpt    5   9832.578  9829.398 ops/ms
IndexInRangeBenchmark.longIndexInRange       259  thrpt    5   4979.326  4947.166 ops/ms
IndexInRangeBenchmark.longIndexInRange       515  thrpt    5   2269.131   614.204 ops/ms
IndexInRangeBenchmark.shortIndexInRange      131  thrpt    5  19865.866 19297.628 ops/ms
IndexInRangeBenchmark.shortIndexInRange      259  thrpt    5  14005.214 13592.407 ops/ms
IndexInRangeBenchmark.shortIndexInRange      515  thrpt    5   8766.450  8531.675 ops/ms
```
As a conclusion, I prefer to keep the current version. WDYT?

jatin-bhateja replied:
Yes, I agrees, I did collect performance with your benchmarks at various AVX levels and see significant gains.

XiaohongGong replied:
Thanks for the performance testing on x86 systems! Agree that a separate PR is fine, and I will address it once this PR merged.

XiaohongGong commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/AbstractMask.java:
```diff
@@ -195,24 +195,52 @@ static long toLongHelper(boolean[] bits) {
+        } else if (limit - offset >= length()) {
+            return this;
+        }
+        return this.and(indexInRange0(offset, limit));
```
Agree! Maybe we could create another PR to handle these optimization in JIT-compiler.

jatin-bhateja reviewed (approved):
LGTM

merykitty commented:
Can you add benchmarks which perform `indexInRange` on really short arrays, such as with size smaller than that of the vectors.

    var m = ispecies.indexInRange(0, size);
    IntVector.fromArray(ispecies, intIn, 0, m).intoArray(intOut, 0, m);

For longer arrays, I think this code shape is not much more verbose and might perform better:

    int bound = ispecies.loopBound(size);
    for (int i = 0; i < bound; i += ispecies.length()) {
        IntVector.fromArray(ispecies, intIn, i).intoArray(intOut, i);
    }
    var m = ispecies.indexInRange(bound, size);
    IntVector.fromArray(ispecies, intIn, bound, m).intoArray(intOut, bound, m);

Thanks.

XiaohongGong commented:
> 

Thanks for the suggestion! Totally agree with you that the tail loop is better to be separated from the main vector loop. Adding smaller value in the ` "@Param" ` set for `size` can match what you mean. I will run the performance with smaller size. Thanks!

XiaohongGong commented:
Hi @merykitty ,

I added a smaller array size (i.e. 7) in the benchmark tests. And to make the test size gap not so large, I changed the original 1027 to 259. 

Following is the benchmark result compared with the jdk mainline, ran on a SVE 512-bit vector size hardware:

```
Benchmark                                 (size) Mode Cnt   Before    After    Units
IndexInRangeBenchmark.byteIndexInRange       7  thrpt  5  35730.865 41386.360  ops/ms
IndexInRangeBenchmark.byteIndexInRange     259  thrpt  5   1902.071  3526.831  ops/ms
IndexInRangeBenchmark.doubleIndexInRange     7  thrpt  5  23329.810 25691.336  ops/ms
IndexInRangeBenchmark.doubleIndexInRange   259  thrpt  5    275.182  1613.951  ops/ms
IndexInRangeBenchmark.floatIndexInRange      7  thrpt  5  24095.898 26462.455  ops/ms
IndexInRangeBenchmark.floatIndexInRange    259  thrpt  5    556.816  2121.836  ops/ms
IndexInRangeBenchmark.intIndexInRange        7  thrpt  5  35722.428 39673.349  ops/ms
IndexInRangeBenchmark.intIndexInRange      259  thrpt  5   4869.695  5368.887  ops/ms
IndexInRangeBenchmark.longIndexInRange       7  thrpt  5  35985.543 41213.203  ops/ms
IndexInRangeBenchmark.longIndexInRange     259  thrpt  5   2303.811  3236.426  ops/ms
IndexInRangeBenchmark.shortIndexInRange      7  thrpt  5  35725.591 41008.665  ops/ms
IndexInRangeBenchmark.shortIndexInRange    259  thrpt  5   1138.442  3172.590  ops/ms
```

Similar improvement can also be observed on X86 AVX-512 machine with -XX:UseAVX=3.

On a 512-bit vector size system, the vector length for long/double is `8`, which is larger than `8` for other types. To make sure the tests just contain the tail loop part, the smaller array size is set to `7`.  Is it ok for you?  It's not meaningful to me that running the smaller size on non-predicated architectures, because the vector boxing is needed because the masked `fromArray()/intoArray()` cannot be vectorized, which can influence the benchmark results.

Thanks,
Xiaohong

merykitty reviewed (approved):
Thanks, the patch LGTM.

XiaohongGong commented:
Thanks for the review @jatin-bhateja @merykitty !

PaulSandoz commented:
I think it would be useful to adjust the naming and comments of some methods to make it clearer the method parameter constraints.

`indexInRange0Helper` is now called if the index is partially or totally out of range at the lower or upper ends and `indexInRange0` is called if partially or totally out of range at the upper end.
e.g. a more literal naming could be:
`AbstractMask::indexInRange0Helper` -> `AbstractMask::indexPartiallyInRangeHelper`
`VectorSupport::indexInRange` -> VectorSupport::indexPartiallyInUpperRange`
?

IIUC the performance numbers show that when the array is not a multiple of the vector size there is still quite an impact overall to calling `VectorSupport::indexInRange` for the last loop iteration. I am guessing the overall loop shape is different which impacts other optimizations?

To do this more optimally likely requires a loop transformation where the last loop iteration is peeled off, but that's a harder transformation in one of the more complicated areas of C2 (although it already supports pre/post loop, so maybe its possible to leverage that?).

XiaohongGong commented:
Thanks for looking at this PR @PaulSandoz !

> I think it would be useful to adjust the naming and comments of some methods to make it clearer the method parameter constraints.
> 
> `indexInRange0Helper` is now called if the index is partially or totally out of range at the lower or upper ends and `indexInRange0` is called if partially or totally out of range at the upper end. e.g. a more literal naming could be: `AbstractMask::indexInRange0Helper` -> `AbstractMask::indexPartiallyInRangeHelper` `VectorSupport::indexInRange` -> VectorSupport::indexPartiallyInUpperRange` ?

The renaming looks good to me. Thanks!

> IIUC the performance numbers show that when the array is not a multiple of the vector size there is still quite an impact overall to calling `VectorSupport::indexInRange` for the last loop iteration. I am guessing the overall loop shape is different which impacts other optimizations?

I think the main influence of the benchmark result comes from the masked ` fromArray()/intoArray()` APIs, especially the masked intoArray() API. For the tail loop part, there is the vector boxing needed on all architectures, with the following reasons:

- If the architecture doesn't support predicate feature, it cannot be intrinsified.
- The `checkMaskFromIndexSize` called inside the `else->if` branch may not be inlined, and the `indexInRange()` generated mask `m` needs the boxing before it.

```
public final
    void intoArray(double[] a, int offset,
                   VectorMask<Double> m) {
        if (m.allTrue()) {
            intoArray(a, offset);
        } else {
            DoubleSpecies vsp = vspecies();
            if (!VectorIntrinsics.indexInRange(offset, vsp.length(), a.length)) {
                checkMaskFromIndexSize(offset, vsp, m, 1, a.length);
            }
            intoArray0(a, offset, m);
        }
    }
```

If the array size is aligned with the vector size, the generated `m` is all true. Hence, the non-masked `intoArray()` is called instead, which improves the performance a lot.

Regarding to the `indexInRange()` API implementation, if the array size is the multiple num of vector size, the branch for the tail loop part is optimized out to an uncommon-trap by C2 compiler, which may improves the performance as well. 

Regarding to the added benchmark, since it is a testing for `indexInRange`, maybe we can remove the calling to the masked `fromArray()/intoArray()` APIs and directly save the mask into a boolean array instead. I guess this may reduce the overall performance gap. 

> 
> To do this more optimally likely requires a loop transformation where the last loop iteration is peeled off, but that's a harder transformation in one of the more complicated areas of C2 (although it already supports pre/post loop, so maybe its possible to leverage that?).

Yes, it is!

XiaohongGong commented:
Hi Paul,

I updated the API name as you suggested and simply the benchmarks by removing the calling to masked `fromArray()/intoArray()` APIs.

Here are the benchmark result compared with jdk/master with ARM NEON:
```
Benchmark                                (size)  Mode  Cnt    Before    After    Units
IndexInRangeBenchmark.byteIndexInRange       7  thrpt   5  164957.447 188954.757 ops/ms
IndexInRangeBenchmark.byteIndexInRange     256  thrpt   5   28373.131  60895.091 ops/ms
IndexInRangeBenchmark.byteIndexInRange     259  thrpt   5   28290.365  55573.807 ops/ms
IndexInRangeBenchmark.byteIndexInRange     512  thrpt   5   15695.618  49147.370 ops/ms
IndexInRangeBenchmark.doubleIndexInRange     7  thrpt   5   58926.711  87837.117 ops/ms
IndexInRangeBenchmark.doubleIndexInRange   256  thrpt   5    2558.505  17795.100 ops/ms
IndexInRangeBenchmark.doubleIndexInRange   259  thrpt   5    2521.995   5309.487 ops/ms
IndexInRangeBenchmark.doubleIndexInRange   512  thrpt   5    1289.556   8882.959 ops/ms
IndexInRangeBenchmark.floatIndexInRange      7  thrpt   5  113429.518 114530.506 ops/ms
IndexInRangeBenchmark.floatIndexInRange    256  thrpt   5    5681.129  31686.156 ops/ms
IndexInRangeBenchmark.floatIndexInRange    259  thrpt   5    5614.762  13659.272 ops/ms
IndexInRangeBenchmark.floatIndexInRange    512  thrpt   5    2897.391  17796.357 ops/ms
IndexInRangeBenchmark.intIndexInRange        7  thrpt   5   50990.391 125139.575 ops/ms
IndexInRangeBenchmark.intIndexInRange      256  thrpt   5    8444.632  31090.867 ops/ms
IndexInRangeBenchmark.intIndexInRange      259  thrpt   5    8349.075  20258.705 ops/ms
IndexInRangeBenchmark.intIndexInRange      512  thrpt   5    4525.218  17555.370 ops/ms
IndexInRangeBenchmark.longIndexInRange       7  thrpt   5   77003.438  89592.650 ops/ms
IndexInRangeBenchmark.longIndexInRange     256  thrpt   5    3669.537  17455.742 ops/ms
IndexInRangeBenchmark.longIndexInRange     259  thrpt   5    3672.086  11150.989 ops/ms
IndexInRangeBenchmark.longIndexInRange     512  thrpt   5    1883.831   8832.311 ops/ms
IndexInRangeBenchmark.shortIndexInRange      7  thrpt   5  159881.634 185593.426 ops/ms
IndexInRangeBenchmark.shortIndexInRange    256  thrpt   5   16762.736  50486.836 ops/ms
IndexInRangeBenchmark.shortIndexInRange    259  thrpt   5   16490.397  35110.418 ops/ms
IndexInRangeBenchmark.shortIndexInRange    512  thrpt   5    8815.322  31113.907 ops/ms
```
And the result with SVE 512-bit vector size:
```
Benchmark                                (size)  Mode  Cnt   Before    After     Units
IndexInRangeBenchmark.byteIndexInRange       7  thrpt   5  48977.004  62712.3874 ops/ms
IndexInRangeBenchmark.byteIndexInRange     256  thrpt   5  28005.444  36067.6281 ops/ms
IndexInRangeBenchmark.byteIndexInRange     259  thrpt   5  26833.661  33337.5660 ops/ms
IndexInRangeBenchmark.byteIndexInRange     512  thrpt   5  18621.850  26251.4372 ops/ms
IndexInRangeBenchmark.doubleIndexInRange     7  thrpt   5  31556.967  63184.8951 ops/ms
IndexInRangeBenchmark.doubleIndexInRange   256  thrpt   5   4394.624  22536.9730 ops/ms
IndexInRangeBenchmark.doubleIndexInRange   259  thrpt   5   4390.727  13714.7822 ops/ms
IndexInRangeBenchmark.doubleIndexInRange   512  thrpt   5   2358.633  15654.2022 ops/ms
IndexInRangeBenchmark.floatIndexInRange      7  thrpt   5  31507.582  62985.8334 ops/ms
IndexInRangeBenchmark.floatIndexInRange    256  thrpt   5   7873.270  25331.0291 ops/ms
IndexInRangeBenchmark.floatIndexInRange    259  thrpt   5   7733.960  22011.2921 ops/ms
IndexInRangeBenchmark.floatIndexInRange    512  thrpt   5   4392.090  21542.3555 ops/ms
IndexInRangeBenchmark.intIndexInRange        7  thrpt   5  55291.415  62846.4699 ops/ms
IndexInRangeBenchmark.intIndexInRange      256  thrpt   5  12580.224  25637.0236 ops/ms
IndexInRangeBenchmark.intIndexInRange      259  thrpt   5  12815.614  23283.9921 ops/ms
IndexInRangeBenchmark.intIndexInRange      512  thrpt   5   7737.667  21611.9642 ops/ms
IndexInRangeBenchmark.longIndexInRange       7  thrpt   5  46632.264  63072.6243 ops/ms
IndexInRangeBenchmark.longIndexInRange     256  thrpt   5   6664.042  22541.1474 ops/ms
IndexInRangeBenchmark.longIndexInRange     259  thrpt   5   6294.857  16994.0206 ops/ms
IndexInRangeBenchmark.longIndexInRange     512  thrpt   5   3446.688  15689.5675 ops/ms
IndexInRangeBenchmark.shortIndexInRange      7  thrpt   5  43243.398  63971.3060 ops/ms
IndexInRangeBenchmark.shortIndexInRange    256  thrpt   5  17997.651  27081.8088 ops/ms
IndexInRangeBenchmark.shortIndexInRange    259  thrpt   5  16572.132  30804.5928 ops/ms
IndexInRangeBenchmark.shortIndexInRange    512  thrpt   5  10211.183  21771.9652 ops/ms
```

Similar gains can also be observed on x86 different systems. From the result, we can see the performance doesn't have too much gap between the 256/259 array sizes.

XiaohongGong commented:
Hi @PaulSandoz ，all your comments have been addressed. Could you please take a look at it again? Thanks a lot!

PaulSandoz reviewed (approved):
Looks good.

XiaohongGong commented:
Thanks for the review @PaulSandoz !

XiaohongGong commented:
/integrate
